src/app/apps/page.tsx

```diff
@@ -1,39 +1,48 @@
+"use client";
+
 import { AppList } from "@/components/AppList";
+import { SessionProvider, useSession } from "next-auth/react";
 import { LoginButton } from "@/components/LoginButton";
-import { getAppData } from "../actions";
+import { useEffect, useState } from "react";
+import {
+  getAppsResponse,
+  getCategoriesResponse,
+  getDevicesResponse,
+} from "@/badgehub-api-client/generated/swagger/public/public";
+import { getAppData, setToken } from "../actions";
 
 export interface SearchParams {
   category: string;
   device: string;
 }
 
-export default async function Listing({
+export default function Listing({
   searchParams,
 }: {
   searchParams: Partial<SearchParams>;
 }) {
-  let data;
-  try {
-    // TODO add caching
-    data = await getAppData(searchParams);
-  } catch (e) {
-    if (!(e instanceof Error)) {
-      return <p>Caught object that wasn&amp;t an error.</p>;
+  const { data: session } = useSession();
+  const [data, setData] = useState<
+    [getAppsResponse, getCategoriesResponse, getDevicesResponse] | null
+  >(null);
+
+  useEffect(() => {
+    async function getData() {
+      const token = (session as any)?.accessToken;
```

Comment:
Maybe better that the session is forwarded as a parameter in getAppData. That way it will be easier to follow.
Now it would be a global var which is possibly going to be set in multiple places. I feel like it's gonna become hard to follow.

Reply:
Code is much better now :)